src/librustc_typeck/check/mod.rs

```diff
@@ -1106,25 +1106,22 @@ fn check_fn<'a, 'gcx, 'tcx>(inherited: &'a Inherited<'a, 'gcx, 'tcx>,
     }
     fcx.demand_suptype(span, ret_ty, actual_return_ty);
 
-    if fcx.tcx.features().termination_trait {
-        // If the termination trait language item is activated, check that the main return type
-        // implements the termination trait.
-        if let Some(term_id) = fcx.tcx.lang_items().termination() {
-            if let Some((id, _)) = *fcx.tcx.sess.entry_fn.borrow() {
-                if id == fn_id {
-                    match fcx.sess().entry_type.get() {
-                        Some(config::EntryMain) => {
-                            let substs = fcx.tcx.mk_substs(iter::once(Kind::from(ret_ty)));
-                            let trait_ref = ty::TraitRef::new(term_id, substs);
-                            let cause = traits::ObligationCause::new(
-                                span, fn_id, ObligationCauseCode::MainFunctionType);
-
-                            inherited.register_predicate(
-                                traits::Obligation::new(
-                                    cause, param_env, trait_ref.to_predicate()));
-                        },
-                        _ => {},
-                    }
+    // Check that the main return type implements the termination trait.
+    if let Some(term_id) = fcx.tcx.lang_items().termination() {
+        if let Some((id, _)) = *fcx.tcx.sess.entry_fn.borrow() {
+            if id == fn_id {
+                match fcx.sess().entry_type.get() {
+                    Some(config::EntryMain) => {
+                        let substs = fcx.tcx.mk_substs(iter::once(Kind::from(ret_ty)));
+                        let trait_ref = ty::TraitRef::new(term_id, substs);
+                        let cause = traits::ObligationCause::new(
+                            span, fn_id, ObligationCauseCode::MainFunctionType);
```

Comment:
the span is specified here; if you look up to line 1028 you can see it comes from the body:

https://github.com/rust-lang/rust/blob/5ccf3ffab24a4237e467b884af6783b1af54bdd8/src/librustc_typeck/check/mod.rs#L1028

In general, to get a good span, you want to look to [the HIR](https://rust-lang-nursery.github.io/rustc-guide/hir.html), which corresponds pretty closely to the input program. In this case, we want to look at the argument `decl`:

https://github.com/rust-lang/rust/blob/5ccf3ffab24a4237e467b884af6783b1af54bdd8/src/librustc_typeck/check/mod.rs#L1001

The `FnDecl` struct is declared here:

https://github.com/rust-lang/rust/blob/5ccf3ffab24a4237e467b884af6783b1af54bdd8/src/librustc/hir/mod.rs#L1718-L1727

We want to get the span from the return type, for which there is a convenient method:

https://github.com/rust-lang/rust/blob/5ccf3ffab24a4237e467b884af6783b1af54bdd8/src/librustc/hir/mod.rs#L1815-L1820

So basically the span should be something like:

```rust
let return_ty_span = decl.output.span();
```

and use that instead of `span`.